### ISSUE TITLE
Sharing-JDBC BLOCKED

### ISSUE THREAD

**Issue by zt9788** [elastic/apm-agent-java#1612]
Sometime http thread will blocked by jdbc 
My programe has blocked for 500 threads
There is no such problem before set APM

springboot2.2
sharding-jdbc3.1.0
mysql-connecter 8.0.18
apm-1.20.0
(apm-server 7.8.1==>kafka)

## Describe the bug
![image](https://user-images.githubusercontent.com/15647179/104292270-977ae100-54f7-11eb-99a4-bc3b1ecd1ac5.png)

"http-nio-9090-exec-512" #1069 daemon prio=5 os_prio=0 tid=0x00007f12b27a5000 nid=0x142d waiting for monitor entry [0x00007f126d4bf000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at com.mysql.cj.jdbc.StatementImpl.getResultSetType(StatementImpl.java:1715)
        - waiting to lock <0x000000074ba2fd08> (a com.mysql.cj.jdbc.ConnectionImpl)
        at com.mysql.cj.jdbc.result.ResultSetFactory.<init>(ResultSetFactory.java:61)
        at com.mysql.cj.jdbc.StatementImpl.<init>(StatementImpl.java:258)
        at com.mysql.cj.jdbc.ConnectionImpl.createStatement(ConnectionImpl.java:1090)
        at com.mysql.cj.jdbc.ConnectionImpl.createStatement(ConnectionImpl.java:1084)
        at com.mysql.cj.jdbc.ConnectionImpl.getMetadataSafeStatement(ConnectionImpl.java:1187)
        at com.mysql.cj.jdbc.ConnectionImpl.getMetadataSafeStatement(ConnectionImpl.java:1183)
        at com.mysql.cj.jdbc.DatabaseMetaData.getUserName(DatabaseMetaData.java:4077)
        at co.elastic.apm.agent.jdbc.helper.JdbcHelper.getConnectionMetaData(JdbcHelper.java:158)
        at co.elastic.apm.agent.jdbc.helper.JdbcHelper.createJdbcSpan(JdbcHelper.java:111)
        at co.elastic.apm.agent.jdbc.StatementInstrumentation$ExecutePreparedStatementInstrumentation.onBeforeExecute(StatementInstrumentation.java:326)
        at java.lang.invoke.LambdaForm$DMH/1509514333.invokeStatic_L_L(LambdaForm$DMH)
        at java.lang.invoke.LambdaForm$MH/1613637839.linkToTargetMethod(LambdaForm$MH)
        at io.shardingsphere.shardingjdbc.jdbc.core.statement.ShardingPreparedStatement.execute(ShardingPreparedStatement.java:134)
        at org.apache.ibatis.executor.statement.PreparedStatementHandler.query(PreparedStatementHandler.java:64)

-------------------------------------------------------------------
"http-nio-9090-exec-14" #263 daemon prio=5 os_prio=0 tid=0x00007f12b2be1000 nid=0x48ef waiting for monitor entry [0x00007f1285dbc000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at com.mysql.cj.jdbc.ConnectionImpl.getCatalog(ConnectionImpl.java:1123)
        - waiting to lock <0x000000074ba2fd08> (a com.mysql.cj.jdbc.ConnectionImpl)
        at com.alibaba.druid.pool.DruidPooledConnection.getCatalog(DruidPooledConnection.java:926)
        at com.alibaba.druid.pool.DruidPooledConnection.prepareStatement(DruidPooledConnection.java:411)
        at io.shardingsphere.shardingjdbc.executor.PreparedStatementExecutor.createPreparedStatement(PreparedStatementExecutor.java:90)
        at io.shardingsphere.shardingjdbc.executor.PreparedStatementExecutor.access$100(PreparedStatementExecutor.java:50)
        at io.shardingsphere.shardingjdbc.executor.PreparedStatementExecutor$1.createStatementExecuteUnit(PreparedStatementExecutor.java:82)
        at io.shardingsphere.core.executor.sql.prepare.SQLExecutePrepareTemplate.getSQLExecuteGroup(SQLExecutePrepareTemplate.java:101)
        at io.shardingsphere.core.executor.sql.prepare.SQLExecutePrepareTemplate.getSQLExecuteGroups(SQLExecutePrepareTemplate.java:92)
        at io.shardingsphere.core.executor.sql.prepare.SQLExecutePrepareTemplate.getSynchronizedExecuteUnitGroups(SQLExecutePrepareTemplate.java:67)
        at io.shardingsphere.core.executor.sql.prepare.SQLExecutePrepareTemplate.getExecuteUnitGroups(SQLExecutePrepareTemplate.java:59)
        at io.shardingsphere.shardingjdbc.executor.PreparedStatementExecutor.obtainExecuteGroups(PreparedStatementExecutor.java:73)
        at io.shardingsphere.shardingjdbc.executor.PreparedStatementExecutor.init(PreparedStatementExecutor.java:68)
        at io.shardingsphere.shardingjdbc.jdbc.core.statement.ShardingPreparedStatement.initPreparedStatementExecutor(ShardingPreparedStatement.java:188)
        at io.shardingsphere.shardingjdbc.jdbc.core.statement.ShardingPreparedStatement.execute(ShardingPreparedStatement.java:136)
        at org.apache.ibatis.executor.statement.PreparedStatementHandler.query(PreparedStatementHandler.java:64)

**Comment by SylvainJuge** [comment]
Hi @zt9788, thanks for reporting this !

Do you have the full thread dump available ?

From what you reported, we can see that the agent might be triggering a deadlock when invoking `getConnectionMetaData` method from `createJdbcSpan` method. Here because the *implementation* of `getConnectionMetaData` within the MySQL is locking on a monitor that is already locked when calling it, we get a deadlock.

```
java.lang.Thread.State: BLOCKED (on object monitor)
at com.mysql.cj.jdbc.StatementImpl.getResultSetType(StatementImpl.java:1715)
- waiting to lock <0x000000074ba2fd08> (a com.mysql.cj.jdbc.ConnectionImpl)
at com.mysql.cj.jdbc.result.ResultSetFactory.(ResultSetFactory.java:61)
at com.mysql.cj.jdbc.StatementImpl.(StatementImpl.java:258)
at com.mysql.cj.jdbc.ConnectionImpl.createStatement(ConnectionImpl.java:1090)
at com.mysql.cj.jdbc.ConnectionImpl.createStatement(ConnectionImpl.java:1084)
at com.mysql.cj.jdbc.ConnectionImpl.getMetadataSafeStatement(ConnectionImpl.java:1187)
at com.mysql.cj.jdbc.ConnectionImpl.getMetadataSafeStatement(ConnectionImpl.java:1183)
at com.mysql.cj.jdbc.DatabaseMetaData.getUserName(DatabaseMetaData.java:4077)
at co.elastic.apm.agent.jdbc.helper.JdbcHelper.getConnectionMetaData(JdbcHelper.java:158)
at co.elastic.apm.agent.jdbc.helper.JdbcHelper.createJdbcSpan(JdbcHelper.java:111)
```

A deadlock involves two locks, and here we can see only one here, thus providing the full thread dump would definitely help to investigate this and find a proper solution.

**Comment by zt9788** [comment]
Sorry, I didn't save the dump file at that time. I only saved a few thread stacks 
(except for the HTTP method, the others are the same). At the same time, I have not repeated this question in the past two days.
The "getcatalog" method is just a deadlock without APM stack information

```
"http-nio-9090-exec-14" #263 daemon prio=5 os_prio=0 tid=0x00007f12b2be1000 nid=0x48ef waiting for monitor entry [0x00007f1285dbc000]

   java.lang.Thread.State: BLOCKED (on object monitor)
	at com.mysql.cj.jdbc.ConnectionImpl.getCatalog(ConnectionImpl.java:1123)
	- waiting to lock <0x000000074ba2fd08> (a com.mysql.cj.jdbc.ConnectionImpl)
	at com.alibaba.druid.pool.DruidPooledConnection.getCatalog(DruidPooledConnection.java:926)
	at com.alibaba.druid.pool.DruidPooledConnection.prepareStatement(DruidPooledConnection.java:411)
	at io.shardingsphere.shardingjdbc.executor.PreparedStatementExecutor.createPreparedStatement(PreparedStatementExecutor.java:90)
	at io.shardingsphere.shardingjdbc.executor.PreparedStatementExecutor.access$100(PreparedStatementExecutor.java:50)
	at io.shardingsphere.shardingjdbc.executor.PreparedStatementExecutor$1.createStatementExecuteUnit(PreparedStatementExecutor.java:82)
	at io.shardingsphere.core.executor.sql.prepare.SQLExecutePrepareTemplate.getSQLExecuteGroup(SQLExecutePrepareTemplate.java:101)
	at io.shardingsphere.core.executor.sql.prepare.SQLExecutePrepareTemplate.getSQLExecuteGroups(SQLExecutePrepareTemplate.java:92)
	at io.shardingsphere.core.executor.sql.prepare.SQLExecutePrepareTemplate.getSynchronizedExecuteUnitGroups(SQLExecutePrepareTemplate.java:67)
	at io.shardingsphere.core.executor.sql.prepare.SQLExecutePrepareTemplate.getExecuteUnitGroups(SQLExecutePrepareTemplate.java:59)
	at io.shardingsphere.shardingjdbc.executor.PreparedStatementExecutor.obtainExecuteGroups(PreparedStatementExecutor.java:73)
	at io.shardingsphere.shardingjdbc.executor.PreparedStatementExecutor.init(PreparedStatementExecutor.java:68)
	at io.shardingsphere.shardingjdbc.jdbc.core.statement.ShardingPreparedStatement.initPreparedStatementExecutor(ShardingPreparedStatement.java:188)
	at io.shardingsphere.shardingjdbc.jdbc.core.statement.ShardingPreparedStatement.execute(ShardingPreparedStatement.java:136)
	at org.apache.ibatis.executor.statement.PreparedStatementHandler.query(PreparedStatementHandler.java:64)
	at org.apache.ibatis.executor.statement.RoutingStatementHandler.query(RoutingStatementHandler.java:79)
	at org.apache.ibatis.executor.SimpleExecutor.doQuery(SimpleExecutor.java:63)
	at org.apache.ibatis.executor.BaseExecutor.queryFromDatabase(BaseExecutor.java:324)
	at org.apache.ibatis.executor.BaseExecutor.query(BaseExecutor.java:156)
	at org.apache.ibatis.executor.CachingExecutor.query(CachingExecutor.java:109)
	at org.apache.ibatis.executor.CachingExecutor.query(CachingExecutor.java:83)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:147)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:140)
	at sun.reflect.GeneratedMethodAccessor576.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.mybatis.spring.SqlSessionTemplate$SqlSessionInterceptor.invoke(SqlSessionTemplate.java:426)
	at com.sun.proxy.$Proxy178.selectList(Unknown Source)
	at org.mybatis.spring.SqlSessionTemplate.selectList(SqlSessionTemplate.java:223)
	at org.apache.ibatis.binding.MapperMethod.executeForMany(MapperMethod.java:147)
	at org.apache.ibatis.binding.MapperMethod.execute(MapperMethod.java:80)
	at org.apache.ibatis.binding.MapperProxy.invoke(MapperProxy.java:93)
	at com.sun.proxy.$Proxy179.selectByUid(Unknown Source)
	at :50
	at controller.$$FastClassBySpringCGLIB$$e1db1e4c.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:769)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
	at org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:88)
	at com.hqwx.restapi.config.log.ControllerInterceptor.aroundMethod(ControllerInterceptor.java:79)
	at sun.reflect.GeneratedMethodAccessor154.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:644)
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:633)
	at org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:70)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689)
	at com.hqwx.restapi.controller.al.v2.StudyPlanController$$EnhancerBySpringCGLIB$$61e42a3b.getUserStudySetting(<generated>)
	at sun.reflect.GeneratedMethodAccessor567.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:190)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:106)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:888)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:793)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:634)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:108)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:526)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:861)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1579)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	- locked <0x0000000766b4e8b8> (a org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)

```

--------

full apm stack information

```
"http-nio-9090-exec-511" #1068 daemon prio=5 os_prio=0 tid=0x00007f12b27a3000 nid=0x1409 waiting for monitor entry [0x00007f126d500000]

   java.lang.Thread.State: BLOCKED (on object monitor)
	at com.mysql.cj.jdbc.StatementImpl.getResultSetType(StatementImpl.java:1715)
	- waiting to lock <0x000000074ba2fd08> (a com.mysql.cj.jdbc.ConnectionImpl)
	at com.mysql.cj.jdbc.result.ResultSetFactory.<init>(ResultSetFactory.java:61)
	at com.mysql.cj.jdbc.StatementImpl.<init>(StatementImpl.java:258)
	at com.mysql.cj.jdbc.ConnectionImpl.createStatement(ConnectionImpl.java:1090)
	at com.mysql.cj.jdbc.ConnectionImpl.createStatement(ConnectionImpl.java:1084)
	at com.mysql.cj.jdbc.ConnectionImpl.getMetadataSafeStatement(ConnectionImpl.java:1187)
	at com.mysql.cj.jdbc.ConnectionImpl.getMetadataSafeStatement(ConnectionImpl.java:1183)
	at com.mysql.cj.jdbc.DatabaseMetaData.getUserName(DatabaseMetaData.java:4077)
	at co.elastic.apm.agent.jdbc.helper.JdbcHelper.getConnectionMetaData(JdbcHelper.java:158)
	at co.elastic.apm.agent.jdbc.helper.JdbcHelper.createJdbcSpan(JdbcHelper.java:111)
	at co.elastic.apm.agent.jdbc.StatementInstrumentation$ExecutePreparedStatementInstrumentation.onBeforeExecute(StatementInstrumentation.java:326)
	at java.lang.invoke.LambdaForm$DMH/1509514333.invokeStatic_L_L(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/1613637839.linkToTargetMethod(LambdaForm$MH)
	at io.shardingsphere.shardingjdbc.jdbc.core.statement.ShardingPreparedStatement.execute(ShardingPreparedStatement.java:134)
	at org.apache.ibatis.executor.statement.PreparedStatementHandler.query(PreparedStatementHandler.java:64)
	at org.apache.ibatis.executor.statement.RoutingStatementHandler.query(RoutingStatementHandler.java:79)
	at org.apache.ibatis.executor.SimpleExecutor.doQuery(SimpleExecutor.java:63)
	at org.apache.ibatis.executor.BaseExecutor.queryFromDatabase(BaseExecutor.java:324)
	at org.apache.ibatis.executor.BaseExecutor.query(BaseExecutor.java:156)
	at org.apache.ibatis.executor.CachingExecutor.query(CachingExecutor.java:109)
	at org.apache.ibatis.executor.CachingExecutor.query(CachingExecutor.java:83)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:147)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:140)
	at sun.reflect.GeneratedMethodAccessor576.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.mybatis.spring.SqlSessionTemplate$SqlSessionInterceptor.invoke(SqlSessionTemplate.java:426)
	at com.sun.proxy.$Proxy178.selectList(Unknown Source)
	at org.mybatis.spring.SqlSessionTemplate.selectList(SqlSessionTemplate.java:223)
	at org.apache.ibatis.binding.MapperMethod.executeForMany(MapperMethod.java:147)
	at org.apache.ibatis.binding.MapperMethod.execute(MapperMethod.java:80)
	at org.apache.ibatis.binding.MapperProxy.invoke(MapperProxy.java:93)
	at com.sun.proxy.$Proxy179.selectByUid(Unknown Source)
	at ...(:930)
	at .....$$FastClassBySpringCGLIB$$e1db1e4c.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:769)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
	at org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:88)
	at com.hqwx.restapi.config.log.ControllerInterceptor.aroundMethod(ControllerInterceptor.java:79)
	at sun.reflect.GeneratedMethodAccessor154.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:644)
	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:633)
	at org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:70)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689)
	at com.hqwx.restapi.controller.al.v2.StudyPlanController$$EnhancerBySpringCGLIB$$61e42a3b.getUserStudySetting(<generated>)
	at sun.reflect.GeneratedMethodAccessor567.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:190)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:106)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:888)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:793)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:634)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:108)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:526)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:861)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1579)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	- locked <0x000000070160a348> (a org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by SylvainJuge** [comment]
We really need a full thread dump for further investigation of this potential deadlock.
Would you be OK to wait until it happens again and capture a full thread dump ?

**Comment by zt9788** [comment]
OK,I'll save it

**Comment by eyalkoren** [comment]
@zt9788 did you manage to reproduce the issue and obtain a full thread dump?

**Comment by zt9788** [comment]
@eyalkoren 

I guess Maybe a few times, it may in the middle of the night (so I couldn't catch the stack information) (my project will run some tasks in the middle of the night, and the traffic is relatively large). I saw the apm there is have no data in that time. About an hour to recover record

I am still observing (at the same time, I will moving the methods to other projects)

I suspect that when sharding-jdbc3.x gets metadata, the used default connect causes deadlock due to the Read/Write Splitting.

Sharding-sphere 4.1.1, which I used in another project, does not have this problem
